homeassistant/components/tado/__init__.py

```diff
@@ -107,6 +114,17 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
 
     await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
+    async def add_meter_reading(call: ServiceCall) -> None:
+        """Send meter reading to Tado."""
+        _LOGGER.info("Service call to add_meter_reading")
+        await hass.async_add_executor_job(
+            tadoconnector.set_meter_reading, call.data[ATTR_READING]
+        )
```

Comment:
move it to a separate `services.py`

Reply:
I'm afraid I am not as knowledgeable about the inner workings of HA as you are. Are you able to help out?

Reply:
> I'm afraid I am not as knowledgeable about the inner workings of HA as you are. Are you able to help out?

You can use the [Fastdotcom service](https://github.com/home-assistant/core/blob/6cab4486f7fe38fd4d851b30ffc2ce7bf3f5cd1f/homeassistant/components/fastdotcom/services.py#L32) as an example (you can skip the coordinator part). Basically the idea is that service logic is separated from the rest, including registering.
The TadoConnector, per entry, is stored in `hass.data[DOMAIN][entry.entry_id][DATA]`. Just make sure to pass it through correctly to `services.py`.

Reply:
nothing special HA here, I just think it'll be easier to read and more future proof, if we move that function to a separate file (called `services.py`) and add a separate `async_setup_services` method. Look [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/lametric/__init__.py#L17-L19) for an example

Reply:
> The TadoConnector, per entry, is stored in `hass.data[DOMAIN][entry.entry_id][DATA]`.

Just to verify my understanding; the `entry.entry_id` is a reference to the Tado instance that was added to the config. There will in most cases be only one but can be many.

How do I make sure that the service call is performed only on the Tado instance (`entry.entry_id`) that it is meant for?

Reply:
by making the `config_entry` another input parameter to the service. `easyenergy` also does this.

Reply:
Thanks! I'll have a look at `easyenergy`

Reply:
> > The TadoConnector, per entry, is stored in `hass.data[DOMAIN][entry.entry_id][DATA]`.
> 
> Just to verify my understanding; the `entry.entry_id` is a reference to the Tado instance that was added to the config. There will in most cases be only one but can be many.
> 
> How do I make sure that the service call is performed only on the Tado instance (`entry.entry_id`) that it is meant for?

That's correct. We need to be specific on the entry, whereas we can't have it if people have multiple Tado accounts linekd, all meter readings will be send to one instance. Take [a look here](https://github.com/home-assistant/core/blob/6cab4486f7fe38fd4d851b30ffc2ce7bf3f5cd1f/homeassistant/components/easyenergy/services.py#L92), on how to pass through the entry_id per instance.

Reply:
> please move it to a separate `services.py`

Please resolve conversation at your discretion.

Reply:
just resolve everything you consider it done 🙂

Reply:
I wanted to leave that to you so that you would have the opportunity to validate that it was indeed resolved.